### ISSUE TITLE
RTMP Segmentation fault FOR ARM

### ISSUE THREAD

**Issue by zhanglibo007** [ossrs/srs#1282]
root@NanoPC-T4:/home/pi/workspace/srs/trunk# ./objs/srs -v
2.0.258

root@NanoPC-T4:/home/pi/workspace/srs/trunk# ./objs/srs  -c conf/rtmp.conf 
[2018-11-30 07:47:35.291][trace][12954][0] XCORE-SRS/2.0.258(ZhouGuowen)
[2018-11-30 07:47:35.292][trace][12954][0] config parse complete
[2018-11-30 07:47:35.292][trace][12954][0] write log to console
[2018-11-30 07:47:35.292][trace][12954][0] srs checking config...
[2018-11-30 07:47:35.293][trace][12954][0] detect intranet address: 127.0.0.1, ifname=lo
[2018-11-30 07:47:35.293][trace][12954][0] retrieve local ipv4 ip=192.168.1.111, index=0
[2018-11-30 07:47:35.293][trace][12954][0] detect intranet address: 192.168.1.111, ifname=wlan0
[2018-11-30 07:47:35.293][warn][12954][0][0] stats network use index=0, ip=192.168.1.111
[2018-11-30 07:47:35.293][warn][12954][0][0] stats disk not configed, disk iops disabled.
[2018-11-30 07:47:35.293][trace][12954][0] write log to console
[2018-11-30 07:47:35.293][trace][12954][0] SRS/2.0.258(ZhouGuowen), stable is SRS/2.0release
[2018-11-30 07:47:35.293][trace][12954][0] license: The MIT License (MIT), Copyright (c) 2013-2015 SRS(ossrs)
[2018-11-30 07:47:35.293][trace][12954][0] primary/master: SRS/2.0release
[2018-11-30 07:47:35.293][trace][12954][0] authors: winlin,wenjie.zhao
[2018-11-30 07:47:35.293][trace][12954][0] contributors: winlin<winlin@vip.126.com> wenjie.zhao<740936897@qq.com> xiangcheng.liu<liuxc0116@foxmail.com> naijia.liu<youngcow@youngcow.net> alcoholyi<alcoholyi@qq.com> byteman<wangchen2011@gmail.com> chad.wang<chad.wang.cn@gmail.com> suhetao<suhetao@gmail.com> Johnny<fengjihu@163.com> karthikeyan<keyanmca@gmail.com> StevenLiu<lq@chinaffmpeg.org> zhengfl<zhengfl_1989@126.com> tufang14<breadbean1449@gmail.com> allspace<allspace@gmail.com> niesongsong<nie950@gmail.com> rudeb0t<nimrod@themanxgroup.tw> CallMeNP<np.liamg@gmail.com> synote<synote@qq.com> lovecat<littlefawn@163.com> panda1986<542638787@qq.com> YueHonghui<hongf.yue@hotmail.com> JuntaoLiu<juntliu@gmail.com> 
[2018-11-30 07:47:35.293][trace][12954][0] uname: Linux NanoPC-T4 4.4.138 #23 SMP Fri Aug 17 14:29:02 CST 2018 aarch64 aarch64 aarch64 GNU/Linux
[2018-11-30 07:47:35.294][trace][12954][0] build: 2018-11-30 06:26:35, little-endian
[2018-11-30 07:47:35.294][trace][12954][0] configure:  --arm
[2018-11-30 07:47:35.294][trace][12954][0] features: --prefix=/usr/local/srs --with-hls --with-hds --with-dvr --without-nginx --with-ssl --without-ffmpeg --with-transcode --with-ingest --with-stat --with-http-callback --with-http-server --without-stream-caster --with-http-api --with-librtmp --without-research --without-utest --without-gperf --without-gmc --without-gmp --without-gcp --without-gprof --with-arm-ubuntu12 --without-mips-ubuntu12 --static --log-trace
[2018-11-30 07:47:35.294][trace][12954][0] arm tool chain: cc=arm-linux-gnueabi-gcc gcc=arm-linux-gnueabi-gcc g++=arm-linux-gnueabi-g++ ar=arm-linux-gnueabi-ar ld=arm-linux-gnueabi-ld randlib=arm-linux-gnueabi-ranlib
[2018-11-30 07:47:35.294][trace][12954][0] conf: conf/rtmp.conf, limit: 1000
[2018-11-30 07:47:35.294][trace][12954][0] SRS for arm(build on ubuntu)
[2018-11-30 07:47:35.294][trace][12954][0] writev limits write 1024 iovs a time
[2018-11-30 07:47:35.294][trace][12954][0] cpu is arm, glibc 2.27
[2018-11-30 07:47:35.294][trace][12954][0] check feature rtmp handshake: on
[2018-11-30 07:47:35.294][trace][12954][0] check feature hls: on
[2018-11-30 07:47:35.294][trace][12954][0] check feature hds: on
[2018-11-30 07:47:35.294][trace][12954][0] check feature http callback: on
[2018-11-30 07:47:35.294][trace][12954][0] check feature http api: on
[2018-11-30 07:47:35.294][trace][12954][0] check feature http server: on
[2018-11-30 07:47:35.294][trace][12954][0] check feature http parser: on
[2018-11-30 07:47:35.294][trace][12954][0] check feature dvr: on
[2018-11-30 07:47:35.295][trace][12954][0] check feature transcode: on
[2018-11-30 07:47:35.295][trace][12954][0] check feature ingest: on
[2018-11-30 07:47:35.295][trace][12954][0] check feature system stat: on
[2018-11-30 07:47:35.295][warn][12954][0][0] check feature compile nginx: off
[2018-11-30 07:47:35.295][warn][12954][0][0] check feature compile ffmpeg: off
[2018-11-30 07:47:35.295][warn][12954][0][0] stream caster: off
[2018-11-30 07:47:35.295][trace][12954][0] MR(merged-read): on, @see https://github.com/ossrs/srs/issues/241
[2018-11-30 07:47:35.295][trace][12954][0] MR(merged-read) default 0 sleep 350
[2018-11-30 07:47:35.295][trace][12954][0] MW(merged-write) default sleep 350
[2018-11-30 07:47:35.295][trace][12954][0] read chunk stream cache cid [0, 16)
[2018-11-30 07:47:35.295][trace][12954][0] default gop cache 1, play queue 30s
[2018-11-30 07:47:35.295][trace][12954][0] complex send algorithm enabled.
[2018-11-30 07:47:35.295][warn][12954][0][0] TCP_NODELAY enabled, may hurts performance.
[2018-11-30 07:47:35.295][trace][12954][0] auto guess socket send buffer by merged write
[2018-11-30 07:47:35.295][trace][12954][0] system default latency in ms: mw(0-350) + mr(0-350) + play-queue(0-30000)
[2018-11-30 07:47:35.295][trace][12954][0] http: root mount to ./objs/nginx/html
[2018-11-30 07:47:35.296][trace][12954][0] st_set_eventsys to epoll
[2018-11-30 07:47:35.296][trace][12954][0] st_init success, use epoll
[2018-11-30 07:47:35.296][trace][12954][100] server main cid=100, pid=12954, ppid=12753, asprocess=0
[2018-11-30 07:47:35.296][trace][12954][100] write pid=12954 to ./objs/srs.pid success!
Segmentation fault (core dumped)

root@NanoPC-T4:/home/pi/workspace/srs/trunk# ./objs/srs  -c conf/srs.conf 
[2018-11-30 07:54:55.727][trace][13015][0] XCORE-SRS/2.0.258(ZhouGuowen)
[2018-11-30 07:54:55.728][trace][13015][0] config parse complete
[2018-11-30 07:54:55.728][trace][13015][0] write log to file ./objs/srs.log
[2018-11-30 07:54:55.729][trace][13015][0] you can: tailf ./objs/srs.log
[2018-11-30 07:54:55.729][trace][13015][0] @see: https://github.com/ossrs/srs/wiki/v1_CN_SrsLog
root@NanoPC-T4:/home/pi/workspace/srs/trunk# tail objs/srs.
srs.log   srs.pid   srs.test  
root@NanoPC-T4:/home/pi/workspace/srs/trunk# tail objs/srs.log 
[2018-11-30 07:54:55.732][trace][13015][0] system default latency in ms: mw(0-350) + mr(0-350) + play-queue(0-30000)
[2018-11-30 07:54:55.732][trace][13015][0] http: root mount to ./objs/nginx/html
[2018-11-30 07:54:55.733][trace][13015][0] start deamon mode...
[2018-11-30 07:54:55.733][trace][13016][0] father process exit. ret=0
[2018-11-30 07:54:55.734][trace][13015][0] grandpa process exit.
[2018-11-30 07:54:55.734][trace][13017][0] son(deamon) process running.
[2018-11-30 07:54:55.735][trace][13017][0] st_set_eventsys to epoll
[2018-11-30 07:54:55.735][trace][13017][0] st_init success, use epoll
[2018-11-30 07:54:55.735][trace][13017][100] server main cid=100, pid=13017, ppid=12753, asprocess=0
[2018-11-30 07:54:55.736][trace][13017][100] write pid=13017 to ./objs/srs.pid success!

**Comment by neonxin** [comment]
Here is the patch. Replace the file trunk/3rdparty/patches/1.st.arm.patch with the attachment. Just try it.
[1.st.arm.patch.txt](https://github.com/ossrs/srs/files/2660564/1.st.arm.patch.txt)

**Comment by zhanglibo007** [comment]
very good！

`TRANS_BY_GPT3`

**Comment by jinzequn** [comment]
Awesome !!

**Comment by winlinvip** [comment]
https://github.com/ossrs/state-threads/tree/srs#branch-srs

Merge from [michaeltalyansky](https://github.com/michaeltalyansky/state-threads) and [xzh3836598](https://github.com/ossrs/state-threads/commit/9a17dec8f9c2814d93761665df7c5575a4d2d8a3), support [ARM](https://github.com/ossrs/state-threads/issues/1).

**Comment by ernestgwilsonii** [comment]
Just in case other folks get stuck here like I did!

```
# srs - https://github.com/ossrs/srs
# Start with an official arm32v7 image for Raspberry Pi 3 B+
docker run --name srs -it -p 1935:1935 -p 8085:8085 arm32v7/python:3.7.3-slim-stretch bash

# Update and install requirements
apt-get update && apt-get upgrade -y
apt-get install -y apt-utils build-essential cmake gcc git net-tools pkg-config sudo unzip wget

# Clone srs repo
cd /opt && git clone https://github.com/ossrs/srs && cd srs/trunk

# Fix: Raspberry Pi 3 B+ "Segmentation fault (core dumped)" - Issue: https://github.com/ossrs/srs/issues/1282
wget https://github.com/ossrs/srs/files/2660564/1.st.arm.patch.txt
rm 3rdparty/patches/1.st.arm.patch
mv 1.st.arm.patch.txt 3rdparty/patches/1.st.arm.patch

# Configure, make and install srs
./configure && make
make install

# Review the srs version
/usr/local/srs/objs/srs -v

# Command line start srs server
/usr/local/srs/objs/srs -c /usr/local/srs/conf/rtmp.conf
```

srs runs now
```
root@69d3ebfa8bd4:/opt/srs/trunk# /usr/local/srs/objs/srs -v
2.0.263

root@69d3ebfa8bd4:/opt/srs/trunk# /usr/local/srs/objs/srs -c /usr/local/srs/conf/rtmp.conf 
[2019-05-19 13:49:35.628][trace][28137][0] XCORE-SRS/2.0.263(ZhouGuowen)
[2019-05-19 13:49:35.630][trace][28137][0] config parse complete
[2019-05-19 13:49:35.630][trace][28137][0] write log to console
[2019-05-19 13:49:35.630][trace][28137][0] srs checking config...
[2019-05-19 13:49:35.631][trace][28137][0] detect intranet address: 127.0.0.1, ifname=lo
[2019-05-19 13:49:35.632][trace][28137][0] retrieve local ipv4 ip=172.17.0.2, index=0
[2019-05-19 13:49:35.632][trace][28137][0] detect intranet address: 172.17.0.2, ifname=eth0
[2019-05-19 13:49:35.632][warn][28137][0][0] stats network use index=0, ip=172.17.0.2
[2019-05-19 13:49:35.632][warn][28137][0][0] stats disk not configed, disk iops disabled.
[2019-05-19 13:49:35.632][trace][28137][0] write log to console
[2019-05-19 13:49:35.632][trace][28137][0] SRS/2.0.263(ZhouGuowen), stable is SRS/2.0release
[2019-05-19 13:49:35.632][trace][28137][0] license: The MIT License (MIT), Copyright (c) 2013-2015 SRS(ossrs)
[2019-05-19 13:49:35.633][trace][28137][0] primary/master: SRS/2.0release
[2019-05-19 13:49:35.633][trace][28137][0] authors: winlin,wenjie.zhao
[2019-05-19 13:49:35.633][trace][28137][0] contributors: winlin<winlin@vip.126.com> wenjie.zhao<740936897@qq.com> xiangcheng.liu<liuxc0116@foxmail.com> naijia.liu<youngcow@youngcow.net> alcoholyi<alcoholyi@qq.com> byteman<wangchen2011@gmail.com> chad.wang<chad.wang.cn@gmail.com> suhetao<suhetao@gmail.com> Johnny<fengjihu@163.com> karthikeyan<keyanmca@gmail.com> StevenLiu<lq@chinaffmpeg.org> zhengfl<zhengfl_1989@126.com> tufang14<breadbean1449@gmail.com> allspace<allspace@gmail.com> niesongsong<nie950@gmail.com> rudeb0t<nimrod@themanxgroup.tw> CallMeNP<np.liamg@gmail.com> synote<synote@qq.com> lovecat<littlefawn@163.com> panda1986<542638787@qq.com> YueHonghui<hongf.yue@hotmail.com> JuntaoLiu<juntliu@gmail.com> 
[2019-05-19 13:49:35.633][trace][28137][0] uname: Linux 69d3ebfa8bd4 4.19.34-v7+ #1211 SMP Mon Apr 8 22:56:37 BST 2019 armv7l GNU/Linux
[2019-05-19 13:49:35.633][trace][28137][0] build: 2019-05-19 13:40:47, little-endian
[2019-05-19 13:49:35.633][trace][28137][0] configure: --x86-x64 
[2019-05-19 13:49:35.633][trace][28137][0] features: --prefix=/usr/local/srs --with-hls --with-hds --with-dvr --without-nginx --with-ssl --without-ffmpeg --with-transcode --with-ingest --with-stat --with-http-callback --with-http-server --without-stream-caster --with-http-api --with-librtmp --without-research --with-utest --without-gperf --without-gmc --without-gmp --without-gcp --without-gprof --without-arm-ubuntu12 --without-mips-ubuntu12 --log-trace
[2019-05-19 13:49:35.633][trace][28137][0] conf: /usr/local/srs/conf/rtmp.conf, limit: 1000
[2019-05-19 13:49:35.633][trace][28137][0] writev limits write 1024 iovs a time
[2019-05-19 13:49:35.634][trace][28137][0] cpu is arm, glibc 2.24
[2019-05-19 13:49:35.634][trace][28137][0] check feature rtmp handshake: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature hls: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature hds: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature http callback: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature http api: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature http server: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature http parser: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature dvr: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature transcode: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature ingest: on
[2019-05-19 13:49:35.634][trace][28137][0] check feature system stat: on
[2019-05-19 13:49:35.634][warn][28137][0][0] check feature compile nginx: off
[2019-05-19 13:49:35.634][warn][28137][0][0] check feature compile ffmpeg: off
[2019-05-19 13:49:35.635][warn][28137][0][0] stream caster: off
[2019-05-19 13:49:35.635][trace][28137][0] MR(merged-read): on, @see https://github.com/ossrs/srs/issues/241
[2019-05-19 13:49:35.635][trace][28137][0] MR(merged-read) default 0 sleep 350
[2019-05-19 13:49:35.635][trace][28137][0] MW(merged-write) default sleep 350
[2019-05-19 13:49:35.635][trace][28137][0] read chunk stream cache cid [0, 16)
[2019-05-19 13:49:35.635][trace][28137][0] default gop cache 1, play queue 30s
[2019-05-19 13:49:35.635][trace][28137][0] complex send algorithm enabled.
[2019-05-19 13:49:35.635][warn][28137][0][0] TCP_NODELAY enabled, may hurts performance.
[2019-05-19 13:49:35.635][trace][28137][0] auto guess socket send buffer by merged write
[2019-05-19 13:49:35.635][trace][28137][0] system default latency in ms: mw(0-350) + mr(0-350) + play-queue(0-30000)
[2019-05-19 13:49:35.636][trace][28137][0] http: root mount to ./objs/nginx/html
[2019-05-19 13:49:35.636][trace][28137][0] st_set_eventsys to epoll
[2019-05-19 13:49:35.637][trace][28137][0] st_init success, use epoll
[2019-05-19 13:49:35.637][trace][28137][100] server main cid=100, pid=28137, ppid=1, asprocess=0
[2019-05-19 13:49:35.637][trace][28137][100] write pid=28137 to ./objs/srs.pid success!
[2019-05-19 13:49:35.648][trace][28137][100] RTMP listen at tcp://0.0.0.0:1935, fd=7
[2019-05-19 13:49:35.649][trace][28137][100] signal installed
[2019-05-19 13:49:35.660][trace][28137][100] http: api mount /console to ./objs/nginx/html/console
[2019-05-19 13:49:35.670][trace][28137][100] ingest thread cid=103, current_cid=100
```

**Comment by winlinvip** [comment]
Dup to #1537

**Comment by winlinvip** [comment]
It seems aarch64 is not the same thing as arm

**Comment by winlinvip** [comment]
更新到3.0.84，使用下面的命令编译：

```
./configure --extra-flags='-D__aarch64__' && make
```

对于一般的ARM服务器，可以直接用`./configure && make`，会默认用定义在md.S中的汇编函数。

**Comment by winlinvip** [comment]
@neonxin Thank you very much, I have merged this patch to ST https://github.com/ossrs/state-threads/issues/9 and patched SRS3 by https://github.com/ossrs/srs/commit/ea8111cff5ead60d5773fe29c72a513c66004f00 and https://github.com/ossrs/srs/commit/84f6f3d0511645db434aaff266563b368a96228b

**Comment by winlinvip** [comment]
Also fixed in #1537
